lib/matplotlib/backend_managers.py

```diff
@@ -14,6 +18,197 @@
+
+    canvas : `matplotlib.backend_bases.FigureCanvasBase`
+        The GUI element on which we draw.
+
+    figure : `matplotlib.figure.Figure`
+        The figure that holds the canvas
+
+    toolbar : `matplotlib.backend_bases.ToolbarBase`
+        The toolbar used for interacting with the figure.
+
+    window : `matplotlib.backend_bases.WindowBase`
+        The window that holds the canvas and toolbar.
+
+    num : int
+        The figure number.
+    """
+    def __init__(self, figure, num, **kwargs):
+        super(FigureManager, self).__init__(**kwargs)
+        self._backend_name, self._backend = get_backend()
+
+        self.num = num
+        self.figure = figure
+
+        self._is_gui = hasattr(self._backend, 'Window')
+        if not self._is_gui:
+            self.window = None
+            return
+
+        self._mainloop = self._backend.MainLoop()
+        self.window = self._backend.Window('Figure %d' % num)
+        self.window.mpl_connect('window_destroy_event', self.destroy)
+
+        w = int(self.figure.bbox.width)
+        h = int(self.figure.bbox.height)
+
+        self.window.add_element(self.figure.canvas, 'center')
+
+        self.toolmanager = ToolManager(self.figure)
+        self.toolbar = self._get_toolbar()
+
+        tools.add_tools_to_manager(self.toolmanager)
+        if self.toolbar:
+            tools.add_tools_to_container(self.toolbar)
+            self.statusbar = self._backend.Statusbar(self.toolmanager)
+            h += self.window.add_element(self.statusbar, 'south')
+
+        if self.toolbar is not None:
+            h += self.window.add_element(self.toolbar, 'south')
+
+        self.window.set_default_size(w, h)
+        self._full_screen_flag = False
+
+        if is_interactive():
+            self.window.show()
+
+        def notify_axes_change(fig):
+            'this will be called whenever the current axes is changed'
+            if self.toolmanager is None and self.toolbar is not None:
+                self.toolbar.update()
+        self.figure.add_axobserver(notify_axes_change)
+
+    @property
+    def figure(self):
+        return self._figure
+
+    @figure.setter
+    def figure(self, figure):
+        if hasattr(self, '_figure'):
+            raise NotImplementedError
+
+        if not figure.canvas:
+            self._backend.FigureCanvas(figure, manager=self,
+                                       backend=self.backend)
+        self._figure = figure
+
+    @property
+    def canvas(self):
+        return self._figure.canvas
+
+    def destroy(self, *args):
+        """Called to destroy this FigureManager.
+        """
+
+        # Make sure we run this routine only once for the FigureManager
+        # This ensures the nasty __del__ fix below works.
+        if getattr(self, '_destroying', False) or self._is_gui is False:
+            return
+
+        self._destroying = True
+        self.figure.canvas.destroy()
+        if self.toolbar:
+            self.toolbar.destroy()
```

Comment:
I do think this a bit too complicated for now as this will depend on the backend... i.e. if one added backend specific widgets to the thing that have a different method name to destroy (assuming they have one).

So for now I would rather leave this open like this, leave the class that created the items have the responsibility for cleaning them up.